File: protocol/reference/raven_protocol/address.py
import hashlib
from . import bech32m

ADDRESS_VERSION = 1

def encode(identity_ed_pub: bytes) -> str:
    addr_hash = hashlib.sha256(identity_ed_pub).digest()[:20]
    payload = bytes([ADDRESS_VERSION]) + addr_hash
    return bech32m.encode("rvn", payload)

def decode(addr: str):
    """Returns (addr_hash_20, version) or None."""
    res = bech32m.decode(addr.strip())
    if res is None:
        return None
    hrp, payload = res
    if hrp != "rvn" or len(payload) != 21:
        return None
    return payload[1:], payload[0]

def to_display(addr: str) -> str:
    data = addr.split("1", 1)[1].upper()
    return "rvn1:" + "-".join(data[i:i+4] for i in range(0, len(data), 4))

def from_display(disp: str) -> str:
    body = disp.replace("rvn1:", "").replace("-", "").lower()
    return "rvn1" + body

File: node/crates/raven-core/src/address.rs
//! RavenAddressV1 — Bech32m over `0x01 || SHA-256(ed_pub)[:20]`.

use sha2::{Digest, Sha256};

use crate::bech32m;

pub const ADDRESS_VERSION: u8 = 1;
pub const HRP: &str = "rvn";

pub fn encode_address(identity_ed_pub: &[u8; 32]) -> String {
    let hash = Sha256::digest(identity_ed_pub);
    let mut payload = Vec::with_capacity(21);
    payload.push(ADDRESS_VERSION);
    payload.extend_from_slice(&hash[..20]);
    bech32m::encode(HRP, &payload).expect("bech32m encode")
}

/// Returns `(addr_hash_20, version)` or `None`.
pub fn decode_address(addr: &str) -> Option<(Vec<u8>, u8)> {
    let (hrp, payload) = bech32m::decode(addr.trim())?;
    if hrp != HRP || payload.len() != 21 {
        return None;
    }
    Some((payload[1..].to_vec(), payload[0]))
}

pub fn to_display(addr: &str) -> String {
    let data = addr.split_once('1').map(|(_, d)| d).unwrap_or(addr);
    let upper = data.to_uppercase();
    let groups: Vec<&str> = upper
        .as_bytes()
        .chunks(4)
        .map(|c| std::str::from_utf8(c).unwrap_or(""))
        .collect();
    format!("rvn1:{}", groups.join("-"))
}

pub fn from_display(disp: &str) -> String {
    let body = disp
        .trim()
        .strip_prefix("rvn1:")
        .unwrap_or(disp)
        .replace('-', "")
        .to_lowercase();
    format!("rvn1{body}")
}

#[cfg(test)]
mod tests {
    use super::*;
    use hex::FromHex;

    #[test]
    fn alice_address_matches_vector() {
        let pub_hex = "d75a980182b10ab7d54bfed3c964073a0ee172f3daa62325af021a68f707511a";
        let ed: [u8; 32] = <[u8; 32]>::from_hex(pub_hex).unwrap();
        let addr = encode_address(&ed);
        assert_eq!(addr, "rvn1qysluvwl5922yctzd0u9gpr06gn3k7ldfvecule0");
        let disp = to_display(&addr);
        assert_eq!(
            disp,
            "rvn1:QYSL-UVWL-5922-YCTZ-D0U9-GPR0-6GN3-K7LD-FVEC-ULE0"
        );
        assert_eq!(from_display(&disp), addr);
    }
}
